src/ae_kqueue.c

```diff
@@ -101,62 +108,101 @@ static void aeApiFree(aeEventLoop *eventLoop) {
 
 static int aeApiAddEvent(aeEventLoop *eventLoop, int fd, int mask) {
     aeApiState *state = eventLoop->apidata;
-    struct kevent ke;
-
+    /* Instead of registering events to kqueue one by one, we buffer events and
+     * register them at once along with retrieving pending events in aeApiPoll. */
     if (mask & AE_READABLE) {
-        EV_SET(&ke, fd, EVFILT_READ, EV_ADD, 0, 0, NULL);
-        if (kevent(state->kqfd, &ke, 1, NULL, 0, NULL) == -1) return -1;
+        EV_SET(state->changes + state->nch, fd, EVFILT_READ, EV_ADD, 0, 0, NULL);
+        /* The current changelist is full, register it to kqueue now and
+         * then rewind it to make rooms for follow-up events. */
+        if (++state->nch == EVENT_ARRAY_SIZE(state->changes)) {
+            if (kevent(state->kqfd, state->changes, state->nch, NULL, 0, NULL))
+                return -1;
+            state->nch = 0; /* rewind the changelist. */
+        }
     }
     if (mask & AE_WRITABLE) {
-        EV_SET(&ke, fd, EVFILT_WRITE, EV_ADD, 0, 0, NULL);
-        if (kevent(state->kqfd, &ke, 1, NULL, 0, NULL) == -1) return -1;
+        EV_SET(state->changes + state->nch, fd, EVFILT_WRITE, EV_ADD, 0, 0, NULL);
+        /* The current changelist is full, register it to kqueue now and
+         * then rewind it to make rooms for follow-up events. */
+        if (++state->nch == EVENT_ARRAY_SIZE(state->changes)) {
+            if (kevent(state->kqfd, state->changes, state->nch, NULL, 0, NULL))
+                return -1;
```

Comment:
I'm not quite sure how we're recovering from this error. If we return this, we know some part of kevent failed, but we aren't reseting the state before returning -1, so it will presumably continue failing on other subsequent calls. In the delete pathway we panic, which arguably makes more sense.

Reply:
This function used to return -1 for a single event in changelist, and the behaviors for the callers of `aeCreateFileEvent` vary depending on their needs: some callers later call `aeDeleteEventLoop` to remove the corresponding file descriptor(s), some panic when they got -1, some just don't seem to care and ignore the returned error. We now have two choices: 1) panic anyway as we did in `aeApiDelEvent`, 2) reset the changelist before returning -1. I would prefer the latter because the most common errors that the `kevent` may return are `ENOENT` and `EBADF` and these errors are reported due to invalid or closed file descriptors in the changelist. This might happen when trying to delete events (also it's safe to ignore the errors from deleting invalid events) and it shouldn't happen when trying to add events unless there is a bug. Since we've added the new logic to ensure that we won't call `kevent` with unregistered events in changelist, we can panic if an error occurs in `aeApiDelEvent`. As for `aeApiAddEvent`,  it also makes sense to panic when an error occurs because we expect no errors when no bugs in the code of `valkey`, but I'd like to keep `kqueue` more consistent with other AE backends given that the rest of AEs return -1 instead of panicking in `aeApiAddEvent`.  Does this sound reasonable to you?

Reply:
I don't think 2) makes sense, since it's changing the API contract. If a connection registered the event loop fd and gets an OK, and then later fails to actually apply it, there is really no way for the original connection to learn about that failure. I also think panic makes sense, but that raises the chance of other software bugs causing issues here. I would prefer the panic approach. This seems like a high risk change though, and would like to better understand the performance implications before signing off on the panic approach.

Reply:
Either way looks reasonable to me, but there are always tradeoffs for choices. For 1), it would inform the system immediately but it comes with a high risk of crashing the process (although it shouldn't happen unless there are bugs in our code or some extremely rare cases caused by system faults). For 2), it goes with a more conservative strategy: it prevents the system from crashing and keeps things going as much as possible, but as you said above it could miss some failures. If 1) looks more feasible to you, we can go with that.